sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/BaseAggregateExec.scala

```diff
@@ -92,7 +95,20 @@ trait BaseAggregateExec extends UnaryExecNode with AliasAwareOutputPartitioning
   override def requiredChildDistribution: List[Distribution] = {
     requiredChildDistributionExpressions match {
       case Some(exprs) if exprs.isEmpty => AllTuples :: Nil
-      case Some(exprs) => ClusteredDistribution(exprs) :: Nil
+      case Some(exprs) =>
+        if (isStreaming) {
+          numShufflePartitions match {
+            case Some(parts) =>
+              StatefulOperatorPartitioning.getClusteredDistributionWithBackwardCompatibility(
+                exprs, parts, conf) :: Nil
+
+            case _ =>
+              throw new IllegalStateException("Expected to set the number of partitions before " +
```

Comment:
Maybe we can also add a `require` assertion at class level, e.g. at Line 38

```
require(isStreaming == numShufflePartitions.isDefined, "Expected to set the number of partitions for streaming aggregate")
```

Or we can define only one variable `numStatefulShufflePartitions: Option[Int]` instead of two: `isStreaming` and `numShufflePartitions`?

Reply:
We create a node with numShufflePartitions = None and replace the value in state rule. That said, we can't check the condition before state rule has been performed.

Reply:
Maybe we should use the new error framework to throw exception in newly added code.

Reply:
ok